File: frontend-admin/src/lib/rbac.ts
/**
 * 平台管理端角色 (RBAC) 与权限判定。
 *
 * 契约来源：docs/API/phase-1-contracts.md §7（RBAC 与权限清单）。
 *
 * 设计说明：
 * - AdminRole 四类角色与后端 admin_users.role 枚举 1:1 对齐
 * - Permission 类型直接使用契约中的字符串常量（形如 `scope:resource:action`）
 * - 前端权限来源为 GET /api/v1/admin/me 返回的 permissions 数组，
 *   由 auth-store 统一持有；hasPermission 从传入的权限集合判断
 * - 前端 gating **不构成安全边界**：后端每次调用仍会做二次鉴权
 */

/**
 * 平台管理端角色枚举（与后端 admin_users.role 一一对应）。
 */
export enum AdminRole {
  /** 超级管理员：拥有所有权限 */
  SUPER_ADMIN = "SUPER_ADMIN",
  /** 业务管理员：商家审核、商品审核、订单大盘等 */
  BUSINESS_ADMIN = "BUSINESS_ADMIN",
  /** 客服管理员：售后仲裁、用户投诉处理 */
  CUSTOMER_SERVICE_ADMIN = "CUSTOMER_SERVICE_ADMIN",
  /** 技术管理员：系统配置、日志、权限分配 */
  TECH_ADMIN = "TECH_ADMIN",
  /** Phase 6 客服组长：仲裁 + 派单管理（释放/重派） */
  CUSTOMER_SERVICE_LEAD = "CUSTOMER_SERVICE_LEAD",
  /** Phase 6 普通客服：仅可见待认领 / 自己认领的仲裁单 */
  CUSTOMER_SERVICE_AGENT = "CUSTOMER_SERVICE_AGENT",
}

/**
 * 权限点联合类型。
 *
 * 命名规则：`{scope}:{resource}[:{sub}]:{action}`，scope ∈ {user, merchant, admin}。
 * Phase 1 使用契约 §7.2 权限键；Phase 2 追加类目/品牌/商品审核（契约 §5）。
 *
 * 尚未在契约中出现但骨架页面仍需引用的占位键保留为未来 Phase 声明，
 * Sidebar 会将无匹配权限的项 disabled 化。
 */
export type Permission =
  // ---- 契约 §7.2 Phase 1 权限清单 ----
  | "user:self:read"
  | "user:self:update"
  | "user:merchant_application:submit"
  | "user:merchant_application:withdraw"
  | "user:merchant_application:read"
  | "merchant:self:read"
  | "merchant:shop:update"
  | "admin:self:read"
  | "admin:merchant_application:read"
  | "admin:merchant_application:review"
  | "admin:audit_log:read"
  // ---- Phase 2 契约 §5 新增：类目 / 品牌 / 商品审核 ----
  | "admin:category:manage"
  | "admin:brand:manage"
  | "admin:spu:review"
  | "admin:spu:force_offshelf"
  | "admin:spu:read_all"
  // ---- Phase 3 契约 §5 新增：订单大盘与干预 ----
  | "admin:order:read_all"
  | "admin:order:intervene"
  | "admin:order:add_note"
  // ---- Phase 4 契约 §6 新增：售后仲裁 ----
  | "admin:aftersales:read_all"
  | "admin:aftersales:arbitrate"
  | "admin:aftersales:force_refund"
  | "admin:aftersales:add_note"
  // ---- Phase 5 契约 §10 新增：评价审核 / 举报处理 / 通知 ----
  | "admin:review:moderate"
  | "admin:review_report:handle"
  | "admin:notification:read"
  // ---- Phase 6 契约新增：管理员账号管理 / 权限矩阵 / 售后派单 ----
  | "admin:user:read"
  | "admin:user:manage"
  | "admin:rbac:read"
  | "admin:aftersales:manage"
  // ---- 后续 Phase 预留（暂无对应权限时返回 false，UI 侧 disabled）----
  | "admin:product:review"
  | "admin:order:read"
  | "admin:refund:arbitrate"
  | "admin:rbac:manage";

/** 角色元信息（用于 UI badge、下拉展示） */
export interface AdminRoleMeta {
  role: AdminRole;
  label: string;
  /** 用于 badge 的中性色调 */
  tone: "primary" | "info" | "warning" | "danger";
  description: string;
  /**
   * 该角色**理论上**默认拥有的权限（仅作 UI 文案与文档参考）。
   *
   * 权限的真实来源仍是后端 GET /admin/me 下发的 permissions 数组，
   * 前端不用此字段做鉴权判断（避免与后端矩阵漂移）。
   *
   * 契约 §5 Phase 2 分配：
   * - SUPER_ADMIN         → 全部
   * - BUSINESS_ADMIN      → 商家审核 + 类目/品牌/商品审核 + 强制下架
   * - CUSTOMER_SERVICE_ADMIN → 只读商品（read_all）+ 售后仲裁（Phase 4）
   * - TECH_ADMIN          → 系统 / RBAC / 日志
   *
   * Phase 4 契约 §6 追加：
   * - admin:aftersales:read_all     → SUPER, BUSINESS, CUSTOMER_SERVICE
   * - admin:aftersales:arbitrate    → SUPER, CUSTOMER_SERVICE
   * - admin:aftersales:force_refund → SUPER, CUSTOMER_SERVICE
   * - admin:aftersales:add_note     → SUPER, CUSTOMER_SERVICE
   *
   * Phase 5 契约 §10 追加：
   * - admin:review:moderate         → SUPER, BUSINESS, CUSTOMER_SERVICE
   * - admin:review_report:handle    → SUPER, CUSTOMER_SERVICE
   * - admin:notification:read       → 任何 admin
   *
   * Phase 6 契约追加：
   * - admin:user:read               → SUPER, TECH_ADMIN（只读账号列表）
   * - admin:user:manage             → SUPER（账号增删改）
   * - admin:rbac:read               → SUPER, BUSINESS_ADMIN, TECH_ADMIN
   * - admin:aftersales:manage       → SUPER, CUSTOMER_SERVICE_LEAD（释放/重派）
   * - 新增角色：CUSTOMER_SERVICE_LEAD / CUSTOMER_SERVICE_AGENT
   */
  defaultPermissions: readonly Permission[];
}

export const ADMIN_ROLE_META: Record<AdminRole, AdminRoleMeta> = {
  [AdminRole.SUPER_ADMIN]: {
    role: AdminRole.SUPER_ADMIN,
    label: "超级管理员",
    tone: "danger",
    description: "拥有所有权限，请谨慎操作",
    defaultPermissions: [
      "admin:self:read",
      "admin:merchant_application:read",
      "admin:merchant_application:review",
      "admin:audit_log:read",
      "admin:category:manage",
      "admin:brand:manage",
      "admin:spu:review",
      "admin:spu:force_offshelf",
      "admin:spu:read_all",
      "admin:order:read",
      "admin:order:read_all",
      "admin:order:intervene",
      "admin:order:add_note",
      "admin:aftersales:read_all",
      "admin:aftersales:arbitrate",
      "admin:aftersales:force_refund",
      "admin:aftersales:add_note",
      "admin:review:moderate",
      "admin:review_report:handle",
      "admin:notification:read",
      "admin:refund:arbitrate",
      "admin:user:read",
      "admin:user:manage",
      "admin:rbac:read",
      "admin:aftersales:manage",
    ],
  },
  [AdminRole.BUSINESS_ADMIN]: {
    role: AdminRole.BUSINESS_ADMIN,
    label: "业务管理员",
    tone: "primary",
    description: "商家 / 商品 / 订单业务侧管理",
    defaultPermissions: [
      "admin:self:read",
      "admin:merchant_application:read",
      "admin:merchant_application:review",
      "admin:category:manage",
      "admin:brand:manage",
      "admin:spu:review",
      "admin:spu:force_offshelf",
      "admin:spu:read_all",
      "admin:order:read",
      "admin:order:read_all",
      "admin:aftersales:read_all",
      "admin:review:moderate",
      "admin:notification:read",
    ],
  },
  [AdminRole.CUSTOMER_SERVICE_ADMIN]: {
    role: AdminRole.CUSTOMER_SERVICE_ADMIN,
    label: "客服管理员",
    tone: "info",
    description: "售后仲裁与用户投诉处理",
    defaultPermissions: [
      "admin:self:read",
      "admin:spu:read_all",
      "admin:refund:arbitrate",
      "admin:order:read",
      "admin:order:read_all",
      "admin:order:intervene",
      "admin:order:add_note",
      "admin:aftersales:read_all",
      "admin:aftersales:arbitrate",
      "admin:aftersales:force_refund",
      "admin:aftersales:add_note",
      "admin:review:moderate",
      "admin:review_report:handle",
      "admin:notification:read",
    ],
  },
  [AdminRole.CUSTOMER_SERVICE_LEAD]: {
    role: AdminRole.CUSTOMER_SERVICE_LEAD,
    label: "客服组长",
    tone: "info",
    description: "仲裁 + 售后派单管理（释放 / 重派）",
    defaultPermissions: [
      "admin:self:read",
      "admin:spu:read_all",
      "admin:order:read_all",
      "admin:order:intervene",
      "admin:order:add_note",
      "admin:aftersales:read_all",
      "admin:aftersales:arbitrate",
      "admin:aftersales:force_refund",
      "admin:aftersales:add_note",
      "admin:aftersales:manage",
      "admin:review:moderate",
      "admin:review_report:handle",
      "admin:notification:read",
    ],
  },
  [AdminRole.CUSTOMER_SERVICE_AGENT]: {
    role: AdminRole.CUSTOMER_SERVICE_AGENT,
    label: "普通客服",
    tone: "info",
    description: "仅处理待认领 / 自己认领的仲裁单",
    defaultPermissions: [
      "admin:self:read",
      "admin:spu:read_all",
      "admin:order:read_all",
      "admin:aftersales:read_all",
      "admin:aftersales:arbitrate",
      "admin:aftersales:add_note",
      "admin:notification:read",
    ],
  },
  [AdminRole.TECH_ADMIN]: {
    role: AdminRole.TECH_ADMIN,
    label: "技术管理员",
    tone: "warning",
    description: "系统配置、日志与权限分配",
    defaultPermissions: [
      "admin:self:read",
      "admin:audit_log:read",
      "admin:user:read",
      "admin:rbac:read",
      "admin:notification:read",
    ],
  },
};

/**
 * 判断权限集合中是否包含指定权限。
 *
 * Phase 1 起前端权限完全由后端下发（GET /api/v1/admin/me 返回的
 * permissions 数组），本函数只做集合包含判断，不再依赖前端硬编码矩阵。
 * 这样后端矩阵变更时无需前后端同步发版。
 *
 * @param permissions 当前登录 admin 拥有的权限（来自 auth store），
 *                    未登录时可传 null/undefined/[] 一律返回 false
 * @param permission 待检查的权限键
 */
export function hasPermission(
  permissions: readonly Permission[] | null | undefined,
  permission: Permission,
): boolean {
  if (!permissions || permissions.length === 0) return false;
  return permissions.includes(permission);
}

/**
 * 便捷方法：具备任一权限即可（用于导航项显示）。
 */
export function hasAnyPermission(
  permissions: readonly Permission[] | null | undefined,
  required: readonly Permission[],
): boolean {
  if (!permissions || permissions.length === 0) return false;
  if (required.length === 0) return true;
  return required.some((p) => permissions.includes(p));
}

File: backend/app/core/rbac.py
"""RBAC primitives — Permissions, role definitions, and role→permission map.

Phase 1 uses a hard-coded permission matrix (contract §7.2). Later
phases may migrate to a database-backed role/permission table if the
matrix grows unwieldy.
"""

from __future__ import annotations

import enum

from app.models.admin_user import AdminRole
from app.models.merchant import MerchantRole


class Permission(enum.StrEnum):
    """Fine-grained permission keys.

    Naming: ``{scope}:{resource}[:{sub}]:{action}``.
    """

    # user scope
    USER_SELF_READ = "user:self:read"
    USER_SELF_UPDATE = "user:self:update"
    USER_MERCHANT_APPLICATION_SUBMIT = "user:merchant_application:submit"
    USER_MERCHANT_APPLICATION_WITHDRAW = "user:merchant_application:withdraw"
    USER_MERCHANT_APPLICATION_READ = "user:merchant_application:read"
    USER_CATALOG_READ = "user:catalog:read"
    USER_SPU_READ = "user:spu:read"
    # user scope — Phase 3
    USER_ADDRESS_MANAGE = "user:address:manage"
    USER_CART_MANAGE = "user:cart:manage"
    USER_ORDER_CREATE = "user:order:create"
    USER_ORDER_READ_OWN = "user:order:read_own"
    USER_ORDER_CANCEL_OWN = "user:order:cancel_own"
    USER_ORDER_CONFIRM_RECEIPT = "user:order:confirm_receipt"
    # user scope — Phase 4 · aftersales
    USER_AFTERSALES_CREATE = "user:aftersales:create"
    USER_AFTERSALES_READ_OWN = "user:aftersales:read_own"
    USER_AFTERSALES_CANCEL_OWN = "user:aftersales:cancel_own"
    USER_AFTERSALES_SUBMIT_TRACKING = "user:aftersales:submit_tracking"
    USER_AFTERSALES_CONFIRM_EXCHANGE = "user:aftersales:confirm_exchange"
    USER_AFTERSALES_NUDGE = "user:aftersales:nudge"
    USER_AFTERSALES_APPEAL = "user:aftersales:appeal"
    USER_UPLOAD_PRESIGN = "user:upload:presign"
    # user scope — Phase 5 · reviews / notifications
    USER_REVIEW_CREATE = "user:review:create"
    USER_REVIEW_EDIT_OWN = "user:review:edit_own"
    USER_REVIEW_DELETE_OWN = "user:review:delete_own"
    USER_REVIEW_REPORT = "user:review:report"
    USER_NOTIFICATION_READ_OWN = "user:notification:read_own"

    # merchant scope
    MERCHANT_SELF_READ = "merchant:self:read"
    MERCHANT_SHOP_UPDATE = "merchant:shop:update"
    MERCHANT_SPU_MANAGE = "merchant:spu:manage"
    MERCHANT_SKU_MANAGE = "merchant:sku:manage"
    MERCHANT_INVENTORY_ADJUST = "merchant:inventory:adjust"
    MERCHANT_UPLOAD_PRESIGN = "merchant:upload:presign"
    # merchant scope — Phase 3
    MERCHANT_ORDER_READ_SHOP = "merchant:order:read_shop"
    MERCHANT_ORDER_SHIP = "merchant:order:ship"
    MERCHANT_ORDER_CANCEL_SHOP = "merchant:order:cancel_shop"
    MERCHANT_ORDER_ADD_NOTE = "merchant:order:add_note"
    # merchant scope — Phase 4 · aftersales
    MERCHANT_AFTERSALES_READ_SHOP = "merchant:aftersales:read_shop"
    MERCHANT_AFTERSALES_REVIEW = "merchant:aftersales:review"
    MERCHANT_AFTERSALES_CONFIRM_RECEIVE = "merchant:aftersales:confirm_receive"
    MERCHANT_AFTERSALES_SHIP_EXCHANGE = "merchant:aftersales:ship_exchange"
    MERCHANT_AFTERSALES_ADD_NOTE = "merchant:aftersales:add_note"
    # merchant scope — Phase 5
    MERCHANT_REVIEW_READ_SHOP = "merchant:review:read_shop"
    MERCHANT_REVIEW_REPLY = "merchant:review:reply"
    MERCHANT_NOTIFICATION_READ_SHOP = "merchant:notification:read_shop"
    MERCHANT_SHOP_UPDATE_PROFILE = "merchant:shop:update_profile"

    # admin scope
    ADMIN_SELF_READ = "admin:self:read"
    ADMIN_MERCHANT_APPLICATION_READ = "admin:merchant_application:read"
    ADMIN_MERCHANT_APPLICATION_REVIEW = "admin:merchant_application:review"
    ADMIN_AUDIT_LOG_READ = "admin:audit_log:read"
    ADMIN_CATEGORY_MANAGE = "admin:category:manage"
    ADMIN_BRAND_MANAGE = "admin:brand:manage"
    ADMIN_SPU_REVIEW = "admin:spu:review"
    ADMIN_SPU_FORCE_OFFSHELF = "admin:spu:force_offshelf"
    ADMIN_SPU_READ_ALL = "admin:spu:read_all"
    # admin scope — Phase 3
    ADMIN_ORDER_READ_ALL = "admin:order:read_all"
    ADMIN_ORDER_INTERVENE = "admin:order:intervene"
    ADMIN_ORDER_ADD_NOTE = "admin:order:add_note"
    ADMIN_TASK_RUN = "admin:task:run"
    # admin scope — Phase 4 · aftersales
    ADMIN_AFTERSALES_READ_ALL = "admin:aftersales:read_all"
    ADMIN_AFTERSALES_ARBITRATE = "admin:aftersales:arbitrate"
    ADMIN_AFTERSALES_FORCE_REFUND = "admin:aftersales:force_refund"
    ADMIN_AFTERSALES_ADD_NOTE = "admin:aftersales:add_note"
    # admin scope — Phase 5
    ADMIN_REVIEW_MODERATE = "admin:review:moderate"
    ADMIN_REVIEW_REPORT_HANDLE = "admin:review_report:handle"
    ADMIN_NOTIFICATION_READ = "admin:notification:read"
    # admin scope — Phase 6 · admin account management
    ADMIN_USER_READ = "admin:user:read"
    ADMIN_USER_MANAGE = "admin:user:manage"
    ADMIN_RBAC_READ = "admin:rbac:read"
    ADMIN_AFTERSALES_MANAGE = "admin:aftersales:manage"


# ---------------------------------------------------------------------------
# Baseline permission sets
# ---------------------------------------------------------------------------

# Every authenticated User gets these unconditionally.
USER_BASE_PERMISSIONS: frozenset[Permission] = frozenset(
    {
        Permission.USER_SELF_READ,
        Permission.USER_SELF_UPDATE,
        Permission.USER_MERCHANT_APPLICATION_SUBMIT,
        Permission.USER_MERCHANT_APPLICATION_WITHDRAW,
        Permission.USER_MERCHANT_APPLICATION_READ,
        Permission.USER_CATALOG_READ,
        Permission.USER_SPU_READ,
        Permission.USER_ADDRESS_MANAGE,
        Permission.USER_CART_MANAGE,
        Permission.USER_ORDER_CREATE,
        Permission.USER_ORDER_READ_OWN,
        Permission.USER_ORDER_CANCEL_OWN,
        Permission.USER_ORDER_CONFIRM_RECEIPT,
        # Phase 4
        Permission.USER_AFTERSALES_CREATE,
        Permission.USER_AFTERSALES_READ_OWN,
        Permission.USER_AFTERSALES_CANCEL_OWN,
        Permission.USER_AFTERSALES_SUBMIT_TRACKING,
        Permission.USER_AFTERSALES_CONFIRM_EXCHANGE,
        Permission.USER_AFTERSALES_NUDGE,
        Permission.USER_AFTERSALES_APPEAL,
        Permission.USER_UPLOAD_PRESIGN,
        # Phase 5
        Permission.USER_REVIEW_CREATE,
        Permission.USER_REVIEW_EDIT_OWN,
        Permission.USER_REVIEW_DELETE_OWN,
        Permission.USER_REVIEW_REPORT,
        Permission.USER_NOTIFICATION_READ_OWN,
    }
)


# Merchant permissions keyed by their in-shop role.
MERCHANT_ROLE_PERMISSIONS: dict[MerchantRole, frozenset[Permission]] = {
    MerchantRole.SHOP_OWNER: frozenset(
        {
            Permission.MERCHANT_SELF_READ,
            Permission.MERCHANT_SHOP_UPDATE,
            Permission.MERCHANT_SPU_MANAGE,
            Permission.MERCHANT_SKU_MANAGE,
            Permission.MERCHANT_INVENTORY_ADJUST,
            Permission.MERCHANT_UPLOAD_PRESIGN,
            Permission.MERCHANT_ORDER_READ_SHOP,
            Permission.MERCHANT_ORDER_SHIP,
            Permission.MERCHANT_ORDER_CANCEL_SHOP,
            Permission.MERCHANT_ORDER_ADD_NOTE,
            # Phase 4
            Permission.MERCHANT_AFTERSALES_READ_SHOP,
            Permission.MERCHANT_AFTERSALES_REVIEW,
            Permission.MERCHANT_AFTERSALES_CONFIRM_RECEIVE,
            Permission.MERCHANT_AFTERSALES_SHIP_EXCHANGE,
            Permission.MERCHANT_AFTERSALES_ADD_NOTE,
            # Phase 5
            Permission.MERCHANT_REVIEW_READ_SHOP,
            Permission.MERCHANT_REVIEW_REPLY,
            Permission.MERCHANT_NOTIFICATION_READ_SHOP,
            Permission.MERCHANT_SHOP_UPDATE_PROFILE,
        }
    ),
    MerchantRole.SHOP_OPERATOR: frozenset(
        {
            Permission.MERCHANT_SELF_READ,
            Permission.MERCHANT_SPU_MANAGE,
            Permission.MERCHANT_SKU_MANAGE,
            Permission.MERCHANT_INVENTORY_ADJUST,
            Permission.MERCHANT_UPLOAD_PRESIGN,
            Permission.MERCHANT_ORDER_READ_SHOP,
            Permission.MERCHANT_ORDER_SHIP,
            Permission.MERCHANT_ORDER_ADD_NOTE,
            # Phase 4
            Permission.MERCHANT_AFTERSALES_READ_SHOP,
            Permission.MERCHANT_AFTERSALES_REVIEW,
            Permission.MERCHANT_AFTERSALES_CONFIRM_RECEIVE,
            Permission.MERCHANT_AFTERSALES_SHIP_EXCHANGE,
            Permission.MERCHANT_AFTERSALES_ADD_NOTE,
            # Phase 5
            Permission.MERCHANT_REVIEW_READ_SHOP,
            Permission.MERCHANT_REVIEW_REPLY,
            Permission.MERCHANT_NOTIFICATION_READ_SHOP,
        }
    ),
    MerchantRole.SHOP_SUPPORT: frozenset(
        {
            Permission.MERCHANT_SELF_READ,
            Permission.MERCHANT_ORDER_READ_SHOP,
            Permission.MERCHANT_ORDER_ADD_NOTE,
            # Phase 4 — SHOP_SUPPORT can only read + note, not review / ship
            Permission.MERCHANT_AFTERSALES_READ_SHOP,
            Permission.MERCHANT_AFTERSALES_ADD_NOTE,
            # Phase 5 — read reviews & notifications; no reply
            Permission.MERCHANT_REVIEW_READ_SHOP,
            Permission.MERCHANT_NOTIFICATION_READ_SHOP,
        }
    ),
}


# Admin permissions keyed by role.
ROLE_PERMISSIONS: dict[AdminRole, frozenset[Permission]] = {
    AdminRole.SUPER_ADMIN: frozenset(Permission),
    AdminRole.BUSINESS_ADMIN: frozenset(
        {
            Permission.ADMIN_SELF_READ,
            Permission.ADMIN_MERCHANT_APPLICATION_READ,
            Permission.ADMIN_MERCHANT_APPLICATION_REVIEW,
            Permission.ADMIN_CATEGORY_MANAGE,
            Permission.ADMIN_BRAND_MANAGE,
            Permission.ADMIN_SPU_REVIEW,
            Permission.ADMIN_SPU_FORCE_OFFSHELF,
            Permission.ADMIN_SPU_READ_ALL,
            Permission.ADMIN_ORDER_READ_ALL,
            # Phase 4 — BUSINESS_ADMIN can read but not arbitrate
            Permission.ADMIN_AFTERSALES_READ_ALL,
            # Phase 5
            Permission.ADMIN_REVIEW_MODERATE,
            Permission.ADMIN_NOTIFICATION_READ,
            # Phase 6 — view-only access to the permission matrix
            Permission.ADMIN_RBAC_READ,
        }
    ),
    AdminRole.CUSTOMER_SERVICE_ADMIN: frozenset(
        {
            Permission.ADMIN_SELF_READ,
            Permission.ADMIN_SPU_READ_ALL,
            Permission.ADMIN_ORDER_READ_ALL,
            Permission.ADMIN_ORDER_INTERVENE,
            Permission.ADMIN_ORDER_ADD_NOTE,
            # Phase 4
            Permission.ADMIN_AFTERSALES_READ_ALL,
            Permission.ADMIN_AFTERSALES_ARBITRATE,
            Permission.ADMIN_AFTERSALES_FORCE_REFUND,
            Permission.ADMIN_AFTERSALES_ADD_NOTE,
            # Phase 5
            Permission.ADMIN_REVIEW_MODERATE,
            Permission.ADMIN_REVIEW_REPORT_HANDLE,
            Permission.ADMIN_NOTIFICATION_READ,
        }
    ),
    AdminRole.CUSTOMER_SERVICE_LEAD: frozenset(
        {
            Permission.ADMIN_SELF_READ,
            Permission.ADMIN_SPU_READ_ALL,
            Permission.ADMIN_ORDER_READ_ALL,
            Permission.ADMIN_ORDER_INTERVENE,
            Permission.ADMIN_ORDER_ADD_NOTE,
            # Phase 4
            Permission.ADMIN_AFTERSALES_READ_ALL,
            Permission.ADMIN_AFTERSALES_ARBITRATE,
            Permission.ADMIN_AFTERSALES_FORCE_REFUND,
            Permission.ADMIN_AFTERSALES_ADD_NOTE,
            # Phase 5
            Permission.ADMIN_REVIEW_MODERATE,
            Permission.ADMIN_REVIEW_REPORT_HANDLE,
            Permission.ADMIN_NOTIFICATION_READ,
            # Phase 6 — can release / reassign claimed arbitration cases
            Permission.ADMIN_AFTERSALES_MANAGE,
        }
    ),
    AdminRole.CUSTOMER_SERVICE_AGENT: frozenset(
        {
            Permission.ADMIN_SELF_READ,
            Permission.ADMIN_SPU_READ_ALL,
            Permission.ADMIN_ORDER_READ_ALL,
            # Phase 4 — agent works the arbitration pool only
            Permission.ADMIN_AFTERSALES_READ_ALL,
            Permission.ADMIN_AFTERSALES_ARBITRATE,
            Permission.ADMIN_AFTERSALES_ADD_NOTE,
            # Phase 5
            Permission.ADMIN_NOTIFICATION_READ,
        }
    ),
    AdminRole.TECH_ADMIN: frozenset(
        {
            Permission.ADMIN_SELF_READ,
            Permission.ADMIN_AUDIT_LOG_READ,
            Permission.ADMIN_TASK_RUN,
            # Phase 5 — admins should always see their inbox
            Permission.ADMIN_NOTIFICATION_READ,
            # Phase 6 — read-only visibility of admin accounts
            Permission.ADMIN_USER_READ,
            Permission.ADMIN_RBAC_READ,
        }
    ),
}


class AftersalesScope(enum.StrEnum):
    """Data-scope for aftersales arbitration visibility."""

    ALL = "all"
    OWN_AND_UNCLAIMED = "own_and_unclaimed"


def aftersales_scope_for_admin(role: AdminRole) -> AftersalesScope:
    """Return the aftersales data-scope for a role.

    ``CUSTOMER_SERVICE_AGENT`` only sees cases that are unclaimed or
    claimed by themselves (still in ``admin_arbitrating``); every other
    role sees the whole queue.
    """
    if role == AdminRole.CUSTOMER_SERVICE_AGENT:
        return AftersalesScope.OWN_AND_UNCLAIMED
    return AftersalesScope.ALL


def permissions_for_admin(role: AdminRole) -> frozenset[Permission]:
    """Return the permission set for a given admin role."""
    return ROLE_PERMISSIONS.get(role, frozenset())


def permissions_for_merchant(role: MerchantRole) -> frozenset[Permission]:
    """Return the permission set for a given merchant role."""
    return MERCHANT_ROLE_PERMISSIONS.get(role, frozenset())
